Feat: handle burn reorgs during tenure extends

kantai opened the pull request:
This PR add an end to end integration test for a burn reorg which invalidates a tenure extend's burn view (but not the tenure's election burn block), and updates the endpoints and logics used to mine and validate block proposals impacted by such reorgs.

This PR:
* Updates signer logic to check the stacks-node's tenure info after local state times out (do not consult local state for global acceptance status).
* Updates the stacks-node's `/v3/tenures/tip` endpoint to return highest tip in tenure with canonical burn_view
* Updates proposal validation and mining logic to use that same logic when checking what the parent of a block proposal should be
* Updates the miner to be more defensive with a burn_view check. Failing this check _should_ be unreachable with the other changes to the miner's parent selection logic.

In the event of one of these reorgs, the signer will eventually "time out" their local state and then consult the stacks-node, which at this point will return the appropriate tenure tip.

Figuring out the appropriate tenure tip (i.e., returning from `/v3/tenures/tip`) requires MARF lookups (to check if a given consensus_hash is in the canonical fork). These are more expensive that then previous SQLite-only lookup, but I believe the overhead here is relatively low: in cases where there is no bitcoin fork, there should only be one MARF lookup (because the SQLite query returns the results in descending block order, as soon as it finds a match, it returns). However, I do think it'll be wise to try to perf/benchmark this with a mock-signer before releasing.

wileyj commented:
@moodmosaic

obycode commented on stacks-signer/src/chainstate.rs:
```diff
@@ -573,43 +572,41 @@ impl SortitionsView {
-    /// the signer may have been added to an already-running node.
-    pub fn check_tenure_change_confirms_parent(
-        tenure_change: &TenureChangePayload,
+    /// If we can't look up `tenure_id`, assume `block` is higher.
```
It would be helpful to explain why this is safe. As a reader, it seems like a dangerous assumption.

kantai replied:
Ah, yeah, it's definitely an assumption that its easy to come up with scenarios where its wrong!

So the rationale for why its _safe_ is roughly:

1. The stacks-node won't approve the block if it can't find its parent or parent tenure ID, so if this check fails here, and we submit the proposal to the stacks-node, unless the stacks-node has processed new state, the stacks-node will reject the proposal (or wait until the parent has been processed). 
2. If the stacks-node does get new state allowing it to process the proposal, `tenure_id` here would be checked in the block proposal endpoint. If it still fails to look up the `tenure_id`, it'll reject the block. Otherwise, it performs its own check to get the most recent block from the tenure (so its using its state). 
3. If you are starting a new tenure, `tenure_id` here would be set to the parent tenure. If your signer doesn't have any non-expired local state about that tenure, then whatever state your stacks-node has is the state which should be used, and again, the block proposal endpoint performs a check to make sure its building off of the latest block in the prior tenure.

obycode commented on stackslib/src/chainstate/stacks/db/mod.rs:
```diff
@@ -862,6 +862,13 @@ const CHAINSTATE_SCHEMA_4: &[&str] = &[
     "#,
 ];
 
+const CHAINSTATE_SCHEMA_5: &[&str] = &[
```
Who came up with this naming scheme 😰? Following the existing naming, maybe this should be `NAKAMOTO_CHAINSTATE_SCHEMA_6` and be added in stackslib/src/chainstate/nakamoto/mod.rs instead?

obycode reviewed (commented):
Looks great! Just one more nit comment.

kantai commented:
I'm still seeing a little bit of test flake locally with the new test, but I removed the source of flakiness that triggered on CI. It was due to unrelated behavior: the new test places a tight assertion on the second 3.0 tenure being "normal", but it was possible that the first block commit in 3.0 pointed at the last 2.5 block rather than the first 3.0 tenure. The test now waits for a block commit pointing at the current stacks tip's consensus hash. I added a new kind of global counter to track this, which could be really useful for removing test flake related to tracking block commits.

kantai commented:
(Closed and reopened because the PR wasn't picking up my most recent commit in the branch!)

kantai commented:
Okay! I think the flakiness is now removed from the new test.

obycode reviewed (approved):
Oops - thought I approved this earlier.  👍

kantai commented on stackslib/src/chainstate/nakamoto/mod.rs:
```diff
@@ -289,16 +290,22 @@ lazy_static! {
+pub static NAKAMOTO_CHAINSTATE_SCHEMA_6: &[&str] = &[
+    // schema change is JUST a new index, but the index is on a table
+    //  created by a migration, so don't add the index to the CHAINSTATE_INDEXES
+    r#"UPDATE db_config SET version = "10";"#,
```
Ah-- no, the answer is somewhat worse than that maybe.

`chainstate/nakamoto/mod.rs` and `chainstate/stacks/db/mod.rs` use the same SQLite database, so they share schema versions, even though the schema definitions are in two different files. Schema version "9" is already in `chainstate/stacks/db/mod.rs`. This is, admittedly, not good. I think we should try to refactor our db schema logic so that all our various databases can use the same logic without tons of re-implementations. I like what Hank did for the signerdb in this PR (https://github.com/stacks-network/stacks-core/pull/6116)
